service/src/event/ordering_task.rs

```diff
@@ -175,188 +122,310 @@ impl OrderingTask {
+                    }
+                }
+            }
         }
+        self.new_deliverable.clear();
+        self.cid_map
+            .iter()
+            .any(|(_, ev)| matches!(ev, StreamEvent::Undelivered(_)))
+    }
+
+    async fn order_events(&mut self, pool: &SqlitePool) -> Result<()> {
+        // We collect everything we can into memory and then order things.
+        // If our prev is the init event or already been delivered, we can mark ourselves as deliverable.
+        // If our prev wasn't deliverable yet, we track it and repeat (i.e. follow its prev if we don't have it)
+
+        let mut deliverable_queue = VecDeque::new();
+        let mut undelivered =
+            VecDeque::from_iter(self.cid_map.iter().filter_map(|(cid, ev)| match ev {
+                StreamEvent::Undelivered(meta) => {
+                    assert_eq!(meta.cid, *cid);
+                    Some((meta.cid, meta.prev))
+                }
+                _ => None,
+            }));
+
+        while let Some((cid, prev)) = undelivered.pop_front() {
+            if let Some(prev_event) = self.cid_map.get(&prev) {
+                match prev_event {
+                    StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory and IS deliverable!"
+                        );
+                        deliverable_queue.push_back(cid)
+                    }
+                    StreamEvent::Undelivered(_) => {
+                        trace!(
+                            %prev,
+                            %cid,
+                            "Found event whose prev is already in memory but NOT deliverable."
+                        );
+                        // nothing to do until it arrives on the channel
+                    }
+                }
+
+                continue;
+            }
+
+            let prev_event = StreamEvent::load_by_cid(pool, prev).await?;
+            if let Some(known_prev) = prev_event {
+                match &known_prev {
+                    StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                        deliverable_queue.push_back(cid);
+                    }
+                    StreamEvent::Undelivered(undelivered_ev) => {
+                        // we'll try to follow this back to something deliverable
+                        undelivered.push_back((undelivered_ev.cid, undelivered_ev.prev));
+                    }
+                }
+                self.add_event(known_prev);
+            } else {
+                trace!("Found event that depends on another event we haven't discovered yet");
+            }
+        }
+
+        let mut newly_ready = deliverable_queue.clone();
+        while let Some(cid) = newly_ready.pop_front() {
+            if let Some(now_ready_ev) = self.remove_by_prev_cid(&cid) {
```

Comment:
I'm confused about how the `prev_map` and `cip_map` are supposed to be maintained.  Based on this line, I would think that those maps are only supposed to store events are are undelivered, because here we are removing from them once they become deliverable.  But in `add_event` above, we add to both of those maps events that are already deliverable.  So do those maps store only undelivered events, or do they store all the events in the stream, or something else?

Reply:
we put everything that's "interesting" in them. they are nearly identical sets, except there is no entry for the init event in the prev_map as it has no prev. when we discover events from the db, they are added to the sets. when we hear about events on the channel, they are always added if they're undelivered, and if they're delivered, we ignore them unless we already have the stream in memory.

we then make a copy of all the undelivered events (`cid, prev`) as a queue and iterate through finding the prev event if we can:
- we check in memory using `prev_map` and then fall back to the database.
- if we find a new prev event, we put it in our maps. 
- if it was undelivered, we add it to the queue we're iterating to try to find it's prev. 
- if we find our prev and it's deliverable, we put ourself in the deliverable queue. 

when we call `completed_processing` after all the updates have been committed to the db, we iterate through our cid_map and change everything we just delivered from undelivered to deliverable. if we have nothing left that's undelivered, we're done and dropped from memory, else we stick around and try to reuse the information we previously collected when we discover new events rather than going to disk again.

Reply:
So I wrote that ☝️ and then looked over it and realized that's not what I did anymore and it was a bit of a mess with the weird removal. So now I do keep the two maps in sync except for the init event not being in the prev map and don't remove anything until the all event for the stream are delivered and we drop it. if we get new deliverable events, we won't load up the old history unless we need it. the `prev_map` is used to find something we've unlocked that is a deliverable candidate. I used more mutable state to modify the `cid_map` and track the queue of newly discovered deliverable events and modified some control flow to hopefully make things clearer/better.